File: orchestrator/masagent/audit.py
"""Orchestrator-side audit log.

A hash-chained JSONL log mirroring the Go audit format, so agent reasoning,
scanner runs, PoC attempts and findings are all attributable and tamper-evident.
Written to orchestrator-<engagement>.jsonl alongside the guard's own log.
"""
from __future__ import annotations

import hashlib
import json
import threading
import time
from pathlib import Path


class Audit:
    def __init__(self, path: str, engagement_id: str) -> None:
        self._path = Path(path)
        self._eid = engagement_id
        self._prev = "genesis"
        self._lock = threading.Lock()

    def _hash(self, event: dict) -> str:
        e = dict(event)
        e["hash"] = ""
        return hashlib.sha256(json.dumps(e, sort_keys=True, separators=(",", ":")).encode()).hexdigest()

    def log(self, kind: str, **fields) -> None:
        with self._lock:
            event = {
                "time": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
                "engagement_id": self._eid,
                "kind": kind,
                **fields,
                "prev_hash": self._prev,
                "hash": "",
            }
            event["hash"] = self._hash(event)
            self._prev = event["hash"]
            with self._path.open("a", encoding="utf-8") as f:
                f.write(json.dumps(event) + "\n")

    def agent(self, agent_id: str, action: str, reasoning: str = "", **detail) -> None:
        self.log("agent", actor=agent_id, reason=action, detail={"reasoning": reasoning, **detail})

    def finding(self, finding_id: str, title: str, severity: str, **detail) -> None:
        self.log("finding", target=finding_id, reason=title, detail={"severity": severity, **detail})

File: scopeguard/internal/audit/audit.go
// Package audit provides an append-only, tamper-evident event log per
// engagement. Every host touched, tool run, agent action, scope decision and
// finding is written here with a timestamp. The log is JSON Lines so it streams
// cheaply and each line is independently parseable.
package audit

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"sync"
	"time"
)

// Event is one audit record. PrevHash + Hash form a hash chain so a deleted or
// altered line is detectable after the fact.
type Event struct {
	Time         time.Time      `json:"time"`
	EngagementID string         `json:"engagement_id"`
	Kind         string         `json:"kind"` // scope_decision | request | tool | agent | finding | note
	Actor        string         `json:"actor,omitempty"`
	Target       string         `json:"target,omitempty"`
	Allowed      *bool          `json:"allowed,omitempty"`
	Reason       string         `json:"reason,omitempty"`
	Detail       map[string]any `json:"detail,omitempty"`
	PrevHash     string         `json:"prev_hash"`
	Hash         string         `json:"hash"`
}

// Logger writes events to an io.Writer under a mutex, maintaining the hash chain.
type Logger struct {
	mu           sync.Mutex
	w            io.Writer
	engagementID string
	prev         string
}

// NewLogger writes to w. genesisHash seeds the chain (use "" for a fresh log).
func NewLogger(w io.Writer, engagementID string) *Logger {
	return &Logger{w: w, engagementID: engagementID, prev: "genesis"}
}

// Open appends to (or creates) a log file for the engagement.
func Open(path, engagementID string) (*Logger, io.Closer, error) {
	f, err := os.OpenFile(path, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0o600)
	if err != nil {
		return nil, nil, err
	}
	return NewLogger(f, engagementID), f, nil
}

func hashEvent(e Event) string {
	// Hash the canonical content excluding the Hash field itself.
	e.Hash = ""
	b, _ := json.Marshal(e)
	sum := sha256.Sum256(b)
	return hex.EncodeToString(sum[:])
}

// Log writes one event, filling in time, engagement, and chain hashes.
func (l *Logger) Log(e Event) error {
	l.mu.Lock()
	defer l.mu.Unlock()
	if e.Time.IsZero() {
		e.Time = time.Now().UTC()
	}
	e.EngagementID = l.engagementID
	e.PrevHash = l.prev
	e.Hash = hashEvent(e)
	l.prev = e.Hash
	b, err := json.Marshal(e)
	if err != nil {
		return err
	}
	if _, err := fmt.Fprintln(l.w, string(b)); err != nil {
		return err
	}
	return nil
}

// Decision is a convenience for the most common event: a scope check result.
func (l *Logger) Decision(actor, target string, allowed bool, reason string) error {
	return l.Log(Event{Kind: "scope_decision", Actor: actor, Target: target, Allowed: &allowed, Reason: reason})
}

// Verify re-reads a JSONL audit stream and confirms the hash chain is intact.
// It returns the number of records and the first broken line (0 if intact).
func Verify(r io.Reader) (records int, brokenLine int, err error) {
	dec := json.NewDecoder(r)
	prev := "genesis"
	line := 0
	for dec.More() {
		line++
		var e Event
		if err := dec.Decode(&e); err != nil {
			return records, line, err
		}
		if e.PrevHash != prev {
			return records, line, nil
		}
		if hashEvent(e) != e.Hash {
			return records, line, nil
		}
		prev = e.Hash
		records++
	}
	return records, 0, nil
}
